ducktape/cluster/remoteaccount.py

```diff
@@ -82,40 +205,92 @@ def ssh_command(self, cmd):
         return r
 
     def ssh(self, cmd, allow_fail=False):
+        """Run the given command on the remote host, and block until the command has finished running.
+
+        :param cmd The remote ssh command
+        :param allow_fail If True, ignore nonzero exit status of the remote command, else raise an RemoteCommandError
+
+        :return The exit status of the command.
+        :raise RemoteCommandError If allow_fail is False and the command returns a non-zero exit status, raises
+            RemoteCommandError.
         """
-        Run the specified command on the remote host. If allow_fail is False and
-        the command returns a non-zero exit status, throws
-        subprocess.CalledProcessError. If allow_fail is True, returns the exit
-        status of the command.
-        """
-        return self._ssh_quiet(self.ssh_command(cmd), allow_fail)
+        self._log(logging.DEBUG, "Running ssh command: %s" % cmd)
+
+        client = self.ssh_client
+        stdin, stdout, stderr = client.exec_command(cmd)
+
+        exit_status = stdin.channel.recv_exit_status()
+        try:
+            if not allow_fail and exit_status != 0:
+                raise RemoteCommandError(self, cmd, exit_status, stderr.read())
+        finally:
+            stdin.close()
+            stdout.close()
+            stderr.close()
+
+        return exit_status
 
     def ssh_capture(self, cmd, allow_fail=False, callback=None):
-        '''Runs the command via SSH and captures the output, yielding lines of the output.'''
-        ssh_cmd = self.ssh_command(cmd)
-        proc = subprocess.Popen(ssh_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, close_fds=True)
+        """Run the given command asynchronously via ssh, and return an SSHOutputIter object.
+
+        Does *not* block
+
+        :param cmd The remote ssh command
+        :param allow_fail If True, ignore nonzero exit status of the remote command, else raise an RemoteCommandError
+        :param callback If set, the iterator returns callback(line) for each line of output instead of the raw output
+
+        :return SSHOutputIter object which allows iteration through each line of output.
+        :raise RemoteCommandError If allow_fail is False and the command returns a non-zero exit status, raises
+            RemoteCommandError.
+        """
+        self._log(logging.DEBUG, "Running ssh command: %s" % cmd)
+
+        client = self.ssh_client
+        stdin, stdout, stderr = client.exec_command(cmd)
 
         def output_generator():
-            for line in iter(proc.stdout.readline, ''):
+
+            for line in iter(stdout.readline, ''):
```

Comment:
@granders Not sure how that makes it safe? My point was that you wont hit EOF, so it won't return, but then a big chunk of data on stderr could fill the entire buffer such that the connection can't get anymore stdout data across the channel. So readline won't return and you'll never have a chance to process the stderr data as you are blocked.